### DISCUSSION TITLE
Why headers default cannot be false

### DISCUSSION THREAD

**Discussion by jyxjjj** [symfony/symfony#54328] [Q&A] [answered]
`vendor/symfony/http-foundation/HeaderBag.php`
```
    private function test4(Request $request): void
    {
        dump(
            $request->header('TEST', false),
            $request->input('TEST', false),
            $request->get('TEST', false),
            $request->post('TEST', false),
            $request->json('TEST', false)
        );
    }
```
```
""
false
false
false
false
```
Why headers default cannot be false?

Is it a bug? Because `?string` allows `false` as `empty string` and `true` as `1`

**Comment by alexandre-le-borgne** [comment]
> An HTTP header consists of its case-insensitive name followed by a colon (:), then by its value.

A header is a string. Its value is a part of this string, so it is also a string.

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers

The HeaderBag implementation has retained this logic. Type casting is left to the developer. 

You can take inspiration from :
vendor/symfony/http-foundation/ParameterBag.php:162

    public function getBoolean(string $key, bool $default = false): bool
    {
        return $this->filter($key, $default, \FILTER_VALIDATE_BOOL, ['flags' => \FILTER_REQUIRE_SCALAR]);
    }

**Reply by jyxjjj** [reply]
But it is the default value instead of I really have a false type, if i really received a false type, of course I need convert it, I think default value will not take any affect with real value, if it is a false string, it returns a false string, non of default value's business.

**Reply by stof** [reply]
`false` is not a string. It is a boolean.

**Reply by jyxjjj** [reply]
> `false` is not a string. It is a boolean.

You didn't seem to understand what i meant.
"X-Some-Thing: false"
it returns "false" as string right?
I can convert it to bool false, i know.
But I set the default value to bool `false`, then it should directly return a bool `false`.
Didn't you see my demo above in the main question or I mistook your meaning?
I don't think this two has any connection, how can i explain clearly?
(bool)"false" === true.
Correctly, i should use null instead of false, i just only found this behavior and cannot understand why.

**Reply by stof** [reply, marked as answer]
You are not setting the default to `false` because it cannot be `false` due to the `?string` type of the argument.

As you are not using strict types, you are setting it to an empty string due to the implicit coercion from bool to string (booleans are one of the remaining parts of potentially surprising implicit coercion). In strict-types mode, your code would trigger a TypeError.
